dispatcher/backend/src/routes/auth/oauth2.py

```diff
@@ -61,96 +68,106 @@ def __call__(self):
 
         # check token exists in database and get expire time and user id
-        collection = RefreshTokens()
-        old_token_document = collection.find_one(
-            {"token": old_refresh_token}, {"expire_time": 1, "user_id": 1}
-        )
+        old_token_document = session.execute(
+            sa.select(dbm.Refreshtoken).where(
+                dbm.Refreshtoken.token == old_refresh_token
+            )
+        ).scalar_one_or_none()
         if old_token_document is None:
             raise InvalidGrant("Refresh token is invalid.")
 
         # check token is not expired
-        expire_time = old_token_document["expire_time"]
-        if expire_time < datetime.now():
+        expire_time = old_token_document.expire_time
+        if expire_time < getnow():
             raise InvalidGrant("Refresh token is expired.")
 
         # check user exists
-        user_id = old_token_document["user_id"]
-        user = Users().find_one({"_id": user_id}, {"password_hash": 0})
-        if user is None:
+        orm_user = session.execute(
+            sa.select(dbm.User).where(dbm.User.id == old_token_document.user_id)
+        ).scalar_one_or_none()
+        if orm_user is None:
             raise InvalidGrant("Refresh token is invalid.")
 
         # generate token
         access_token = LoadedAccessToken(
-            user["_id"], user["username"], user.get("scope", {})
+            orm_user.id, orm_user.username, orm_user.scope or {}
         ).encode()
-        refresh_token = OAuth2.generate_refresh_token(user["_id"])
+        refresh_token = OAuth2.generate_refresh_token(orm_user.id, session)
 
         # delete old refresh token from database
-        collection.delete_one({"token": old_refresh_token})
-        collection.delete_many({"expire_time": {"$lte": datetime.now()}})
+        session.delete(old_token_document)
+        session.execute(
+            sa.delete(dbm.Refreshtoken).where(dbm.Refreshtoken.expire_time < getnow())
+        )
 
+        session.flush()
         return OAuth2.success_response(access_token, refresh_token)
 
     @staticmethod
-    def generate_refresh_token(user_id: ObjectId) -> UUID:
+    def generate_refresh_token(user_id: UUID, session: so.Session) -> UUID:
         """Generate and store refresh token in database.
 
         :param user_id: id of user to associate the refresh token with
         :return: a UUID4 refresh token
         """
 
         refresh_token = uuid4()
-        RefreshTokens().insert_one(
-            {
-                "token": refresh_token,
-                "user_id": user_id,
-                "expire_time": datetime.now() + timedelta(days=30),
-            }
+
+        # TODO: fetch "now" from database
+        refresh_token_db = dbm.Refreshtoken(
+            mongo_val=None,
+            mongo_id=None,
+            token=refresh_token,
+            expire_time=getnow() + timedelta(days=30),
         )
+        refresh_token_db.user_id = user_id
```

Comment:
`user_id` is not part of the `__init__` attributes of the dataclass, so that we can create a user and its refresh token in a single run and let the ORM do the job, just like we do for ssh_keys:
```
 orm_user = dbm.User(
            mongo_val=None,
            mongo_id=None,
            username="bob",
            email=None,
            password_hash="somehash",
            scope={},
        )
        session.add(orm_user)
        orm_user.refresh_tokens.append(
            dbm.Refreshtoken(
                ...
            )
        )
```
Is it more self-explanatory if I add the newly created refresh token to the ones of the user instead of just setting the user_id ?

Reply:
Actually I think it would be clearer. Thanks for the explanation and the proposal

Reply:
While implementing / testing this, I realized the current approach is probably the good one indeed.

The current approach is 
```
        refresh_token_db.user_id = user_id
        session.add(refresh_token_db)
```
The alternative would be
```
        user.refresh_tokens.add(refresh_token_db)
```
While the alternative looks simpler (one line instead of two) and more explicit, it has many drawbacks:
- as soon as we access the `refresh_tokens` collection, the ORM is fetching all of them in memory (I confirmed this in manual tests); this would be a significant impact on performance / memory for users with many refresh_tokens
- it is indeed less versatile since we need to have access to the whole ORM user object instead of just its ID, meaning that we cannot optimize SQL by loading only few user properties instead of the whole object when it makes sense

So I would prefer to stay with current code.

Reply:
Indeed, makes little sense to load the whole User just for this specific action. Let's keep it as it is 👍

I was a bit worried about `user.refresh_tokens.add(refresh_token_db)` loading all _refresh_tokens_ in memory. I checked and in fact when building an ORM instance, absolutely everything is loaded immediately. For the user, that includes all the tokens, ssh keys… and the mongo_val. Another reason to get rid of it as soon as the transition is complete.
I'm glad I checked as it's important we understand how the ORM default queries work.